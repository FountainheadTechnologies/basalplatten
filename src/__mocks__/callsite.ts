module.exports = () => ([{
}, {
  getFileName: () => 'test-caller.js'
}]);
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
