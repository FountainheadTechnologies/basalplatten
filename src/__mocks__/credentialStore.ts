export const credentialStore = {
  setToken: jest.fn(),
  getToken: jest.fn(),
  clearToken: jest.fn()
}
