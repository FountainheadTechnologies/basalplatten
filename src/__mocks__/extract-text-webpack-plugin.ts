module.exports = class ExtractTextPlugin {
  extract() {}
};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
