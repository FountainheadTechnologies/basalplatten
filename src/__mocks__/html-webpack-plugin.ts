module.exports = class HtmlWebpackPlugin {};
// @hack: Make module look like es6 to prevent mangling
module.exports.__esModule = true;
