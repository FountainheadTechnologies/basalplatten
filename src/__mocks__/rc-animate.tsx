import React from 'react';

const RcAnimate = props => (
  <div>{props.component}</div>
)

RcAnimate.displayName = 'RcAnimate';

export default RcAnimate;
