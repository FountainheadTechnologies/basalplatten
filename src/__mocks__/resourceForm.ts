const submit = jest.fn();

export const createToken = {
  schema: {
    required: ['username', 'password'],
    properties: {
      username: { type: 'string' },
      password: { type: 'string' }
    },
  },

  submit
}

export const createPerson = {
  schema: {
    required: ['forename', 'surname'],
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      }
    }
  },

  submit
}

export const updatePerson = {
  schema: {
    required: [],
    default: {
      forename: 'Test',
      surname: 'Person',
      date_of_birth: '1985-08-02T23:00:00.000Z'
    },
    properties: {
      forename: { type: 'string' },
      surname: { type: 'string' },
      date_of_birth: {
        type: 'string',
        format: 'date-time'
      }
    }
  }
}
