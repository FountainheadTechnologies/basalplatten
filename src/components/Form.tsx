import React from 'react';
import PropTypes from 'prop-types';
import { Resource, Form as HALForm } from '@optics/hal-client';
import { Data, Schema } from '@optics/hal-client/dist/Form';

import AntForm from 'antd/lib/form';
import 'antd/lib/form/style';

import Button from 'antd/lib/button';
import 'antd/lib/button/style';

import { FormProps as AntFormProps, WrappedFormUtils } from 'antd/lib/form/Form';
import { FormItemProps as AntFormItemProps } from 'antd/lib/form/FormItem';

const DEFAULT_SUBMIT_BUTTON = (
  <Button type="primary">
    Submit
  </Button>
);

export const childContextTypes = {
  schema: PropTypes.object,
  defaultItemProps: PropTypes.object
}

export interface Props {
  resourceForm: HALForm;
  submitButton?: React.ReactElement<any> | false;
  onSuccess?: (result: Resource) => void;
  onFailure?: (error: Error) => void;
  defaultItemProps?: Partial<AntFormItemProps>;
}

export interface State {
  submitting: boolean;
}

export interface Context {
  schema: Schema;
  defaultItemProps: Partial<AntFormItemProps>;
}

class FormComponent extends React.Component<Props & AntFormProps> {
  state: State = {
    submitting: false
  };

  static childContextTypes = childContextTypes;

  render() {
    const {
      resourceForm,
      submitButton,
      onSuccess,
      onFailure,
      children,
      form,
      ...formProps
    } = this.props;

    return (
      <AntForm
        {...formProps}
        onSubmit={this._onSubmit}
      >
        {children}
        {this._renderSubmitButton()}
      </AntForm>
    );
  }

  getChildContext() {
    return {
      schema: this.props.resourceForm.schema,
      defaultItemProps: this.props.defaultItemProps
    }
  }

  protected _renderSubmitButton() {
    const {
      submitButton = DEFAULT_SUBMIT_BUTTON
    } = this.props;

    if (submitButton === false) {
      return;
    }

    return React.cloneElement(submitButton, {
      loading: this.state.submitting,
      htmlType: 'submit'
    });
  }

  protected _validate = () =>
    new Promise<Data>((resolve, reject) => {
      (this.props.form as WrappedFormUtils)
        .validateFields((err, values) => {
          if (err) {
            return reject(Error('One or more fields are invalid'));
          }

          resolve(values);
        });
    })

  protected _onSubmit: React.EventHandler<React.FormEvent<any>> = e => {
    console.log('SUBMIT');
    e.preventDefault();

    this.setState({ submitting: true });

    this._validate()
      .then(values => {
        return this.props.resourceForm.submit(values);
      })
      .then(result => {
        if (this.props.onSuccess) {
          return this.props.onSuccess(result);
        }
      })
      .catch(error => {
        if (this.props.onFailure) {
          return this.props.onFailure(error);
        }

        throw error;
      })
      .finally(() => {
        this.setState({ submitting: false });
      });
  }
}

export const Form = AntForm.create()(FormComponent as any) as React.ComponentClass<Props>;
