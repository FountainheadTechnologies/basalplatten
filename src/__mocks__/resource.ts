import { createToken } from './resourceForm';

export const api = {
  properties: {},
  meta: {},

  config: {
    headers: {}
  },

  formNamed(rel: string, name: string) {
    if (rel === 'token' && name === 'create') {
      return createToken;
    }
  }
}
