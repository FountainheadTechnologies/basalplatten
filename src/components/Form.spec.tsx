import React from 'react';
import promiseFinally from 'promise.prototype.finally';
import { mount, ReactWrapper } from 'enzyme';

import { Form, FormComponent } from './Form';
import { createPerson } from '../__mocks__/resourceForm';

promiseFinally.shim();

const mockEvent = {
  preventDefault: jest.fn()
} as React.Event<any>

describe('default behaviour', () => {
  let form: ReactWrapper;
  let onSuccess = jest.fn();
  let onFailure = jest.fn();

  beforeEach(() => {
    form = mount(
      <Form
        resourceForm={createPerson}
        onSuccess={onSuccess}
        onFailure={onFailure}
      />
    );
  });

  it('renders an empty Ant form with a Submit button', () => {
    expect(form).toMatchSnapshot();
    expect(form.find('Button').exists()).toBe(true);
  });

  describe('when the form is submitted', () => {
    it('calls `props.onSuccess`', async () => {
      const result = { status: 'updated' };
      createPerson.submit.mockReturnValue(Promise.resolve(result));

      await form.find('Form').props().onSubmit(mockEvent);
      expect(onSuccess).toHaveBeenCalledWith(result);
    });

    it('calls `props.onFailure`', async () => {
      const error = Error('oh dear');
      createPerson.submit.mockReturnValue(Promise.reject(error));

      await form.find('Form').props().onSubmit(mockEvent);
      expect(onFailure).toHaveBeenCalledWith(error);
    });
  });

  describe('when form validation fails', () => {
    it('calls `props.onFailure`', async () => {
      form.find('FormComponent').prop('form')
        .validateFields = jest.fn().mockImplementation(cb => {
          cb(Error('User Name is required'), null);
        });

      await form.find('Form').props().onSubmit(mockEvent);
      expect(onFailure).toHaveBeenCalledWith(Error('One or more fields are invalid'));
    });
  });
});

describe('when `props.submitButton` is false', () => {
  let form: ReactWrapper;

  beforeEach(() => {
    form = mount(
      <Form
        resourceForm={createPerson}
        submitButton={false}
      />
    );
  });

  it('omits the Submit button', () => {
    expect(form.find('Button').exists()).toBe(false);
  });
});
