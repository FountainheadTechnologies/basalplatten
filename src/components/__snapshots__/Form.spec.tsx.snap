// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default behaviour renders an empty Ant form with a Submit button 1`] = `
<Form(WrappedComponent)
  onFailure={[Function]}
  onSuccess={[Function]}
  resourceForm={
    Object {
      "schema": Object {
        "properties": Object {
          "date_of_birth": Object {
            "format": "date-time",
            "type": "string",
          },
          "forename": Object {
            "type": "string",
          },
          "surname": Object {
            "type": "string",
          },
        },
        "required": Array [
          "forename",
          "surname",
        ],
      },
      "submit": [Function],
    }
  }
>
  <Component
    form={
      Object {
        "getFieldDecorator": [Function],
        "getFieldError": [Function],
        "getFieldInstance": [Function],
        "getFieldProps": [Function],
        "getFieldValue": [Function],
        "getFieldsError": [Function],
        "getFieldsValue": [Function],
        "isFieldTouched": [Function],
        "isFieldValidating": [Function],
        "isFieldsTouched": [Function],
        "isFieldsValidating": [Function],
        "isSubmitting": [Function],
        "resetFields": [Function],
        "setFields": [Function],
        "setFieldsInitialValue": [Function],
        "setFieldsValue": [Function],
        "submit": [Function],
        "validateFields": [Function],
        "validateFieldsAndScroll": [Function],
      }
    }
    onFailure={[Function]}
    onSuccess={[Function]}
    resourceForm={
      Object {
        "schema": Object {
          "properties": Object {
            "date_of_birth": Object {
              "format": "date-time",
              "type": "string",
            },
            "forename": Object {
              "type": "string",
            },
            "surname": Object {
              "type": "string",
            },
          },
          "required": Array [
            "forename",
            "surname",
          ],
        },
        "submit": [Function],
      }
    }
    wrappedComponentRef={[Function]}
  >
    <FormComponent
      form={
        Object {
          "getFieldDecorator": [Function],
          "getFieldError": [Function],
          "getFieldInstance": [Function],
          "getFieldProps": [Function],
          "getFieldValue": [Function],
          "getFieldsError": [Function],
          "getFieldsValue": [Function],
          "isFieldTouched": [Function],
          "isFieldValidating": [Function],
          "isFieldsTouched": [Function],
          "isFieldsValidating": [Function],
          "isSubmitting": [Function],
          "resetFields": [Function],
          "setFields": [Function],
          "setFieldsInitialValue": [Function],
          "setFieldsValue": [Function],
          "submit": [Function],
          "validateFields": [Function],
          "validateFieldsAndScroll": [Function],
        }
      }
      onFailure={[Function]}
      onSuccess={[Function]}
      resourceForm={
        Object {
          "schema": Object {
            "properties": Object {
              "date_of_birth": Object {
                "format": "date-time",
                "type": "string",
              },
              "forename": Object {
                "type": "string",
              },
              "surname": Object {
                "type": "string",
              },
            },
            "required": Array [
              "forename",
              "surname",
            ],
          },
          "submit": [Function],
        }
      }
      wrappedComponentRef={[Function]}
    >
      <Form
        hideRequiredMark={false}
        layout="horizontal"
        onSubmit={[Function]}
        prefixCls="ant-form"
      >
        <form
          className="ant-form ant-form-horizontal"
          onSubmit={[Function]}
        >
          <Button
            clicked={false}
            ghost={false}
            htmlType="submit"
            loading={false}
            prefixCls="ant-btn"
            type="primary"
          >
            <button
              className="ant-btn ant-btn-primary"
              onClick={[Function]}
              type="submit"
            >
              <span
                key=".0"
              >
                Submit
              </span>
            </button>
          </Button>
        </form>
      </Form>
    </FormComponent>
  </Component>
</Form(WrappedComponent)>
`;
